add support for barcode list

pdimens opened the pull request:
adds:
- option `--barcodes` that lets a user input a barcode file. defaults to the 4M 10X file

modification:
- slight pythonic idiomatic printing changes

fixes #49

pdimens committed:
add input barcode support with validation

pdimens committed:
update version

pdimens committed:
move output to stderr

pdimens commented:
I'm still tweaking this, specifically, I noticed the number of quality scores doesnt yet equal the number of bases. I will confirm when I think it's correct

pdimens commented:
@davidebolo1993 I have a question regarding the process. Given that the GEM chemistry is no longer a thing, it might make more sense for the simulation to favor a simpler design of "one molecule (and its subsequent fragments), one barcode". I'm not 100% clear on the 10X GEM design, but my understanding is that multiple molecules (related? unrelated?) can be in a droplet and all of them get the same barcode? Is that a correct interpretation? If I get a bit of clarity on this, I may be able to augment XENIA to favor this approach.

pdimens committed:
compressionlevel set to 6 (system default, better speed)
removed random 6mer suffix
quality scores correct length

pdimens committed:
fix minor typo

pdimens commented:
Ok, fixed and ready for review

pdimens committed:
set the barcode default and abspath at the argparser level

pdimens committed:
scale droplets to # barcodes
use pysam bgzipper b/c it's faster
status text for barcode validations

pdimens committed:
add error catching when no more barcodes left

pdimens committed:
add support for different output formats

pdimens commented:
Closing this PR temporarily as I'm adding two important features that will take time:
- options for output format (stLFR, etc)
- options for input linked-read style so the simulator can simulate those chemistries (largely a matter of wgsim creating the correct read lengths

pdimens committed:
move assistant functions to own file, add input format options, validations

pdimens committed:
better docstring

pdimens committed:
better docstring

pdimens committed:
changes

pdimens committed:
fix wording

pdimens committed:
fixes

davidebolo1993 commented:
Hi @pdimens  - thanks for all the work. Sure, let me know when it's ready so that I can quickly go through the code. Thanks!
Davide

pdimens committed:
add error handling for too many output barcodes and too-big barcodes

pdimens committed:
better fastq compression text

pdimens commented:
@davidebolo1993 I think it's ready for review. There's been a lot of stuff added to achieve what I hoped to achieve:
- new items in `c` to store important info
- barcodes are stored as a `list(iter(bc))` or a `itertools.product()` generator to leverage `next()` and not need to recursively delete
- gzipping is done by the tabix bgzip module of `pysam` (since it was already imported). It should be a lot faster than native python gzip
- number of droplets scales to the number of total barcodes
  - it was previously 4million, which I inferred reflected the 4M 10X barcodes, so I would like a sanity check if my change is correct
- lots of validations for the new stuff including when a combinatorial barcode will only leave <5bp of insert left after removal
- the BED file was moved as a mandatory arg to the end of the CLI
- options to specify input barcode _type_, which will change the lengths wgsim will simulate for R1/R2
- input barcode type will also change the way the barcodes are parsed
  - 10x/tellseq will use the barcodes as provided
  - haplotagging will do a combinatorial 4-barcode of the ones provided (`itertools.product(bc, bc, bc, bc)`)
  - stlfr will do a combinatorial 3-barcode of the ones provided (`itertools.product(bc, bc, bc)`)
- option to specify the output fastq _type_
  -  e.g. you can have the simulator do an stlfr experiment but output the FASTQ in tellseq format
  - e.g. you can have the simulator do a 10X experiment but output haplotagging format
  - defaults to output type being the same as the barcode type
- to make it generally easier to work on this, I moved most of the XENIA functions to `xenia_helper.py`
- glob pattern includes `.fasta` too
  - would like to have added gz support, but that requires adding BioPython as a dep for pyfaidx
  - if that's something you're cool with, we can do that (I generally try to avoid adding extra non-stdlib deps)
- outputs a barcode whitelist `{output}/whitelist.bc` (more useful for combinatorials)

pdimens committed:
add printing barcodes to whitelist

pdimens commented:
@davidebolo1993 would you be open to moving XENIA into its own project/repo that can be maintained by my team? You (and VISOR) will have full attribution for the original XENIA work in this putative repository (and access rights).

My reasoning is that the proposed changes move XENIA beyond the original scope of a 10X read simulator and it can coexist independently of VISOR and the dependencies of hacks, laser, shorts (pbsim, bwa, minimap, badreads, etc)

pdimens committed:
molecular coverage >= 1 now activates a second mode of operation where coverage is an integer drawn from a normal distribution with mean = molcov and stdv = molcov/3

moves -f --fasta option to positional at the end

adds "standard" output format

adds R1/R2 sequence designations (/1 or 1:N:0:ATAGC) depending on output type

adds -p --prefix file option, defaulting to SIM

pdimens commented:
@davidebolo1993 I added additional features, described below. I suspect you're quite busy, but do you have an idea of when you may be able to review these?

- when `--molecular-coverage` >= 1, the number of reads (N) per molecule becomes an integer drawn from a normal distribution of mean `molcov` and stdev `molcov/3`
  - is capped at 0 and the max N for a molecule
  - very useful for simulating specific sequencing treatments
- moves the `--fasta/-f` option as a positional to the end to simplify the CLI a bit
- added the `standard` output type, which is similar to haplotagging, except the barcode format can be whatever and there is a `BV:i` tag that indicates whether the barcode is valid (`1`) or invalid (`0`)
- added the within-read forward/reverse sequence designations
  - tellseq, 10x, stlfr output uses the modern illumina format e.g. `@SEQID 1:N:0:ATAGCA` or `@SEQID 2:N:0:ATAGCA`
  - haplotagging, standard use the "old" illumina convention `@SEQID/1` or `@SEQID/2`
- add a `--prefix/-p` option to specify an output filename, defaults to the current `SIM`

davidebolo1993 commented:
> @davidebolo1993 would you be open to moving XENIA into its own project/repo that can be maintained by my team? You (and VISOR) will have full attribution for the original XENIA work in this putative repository (and access rights).
> 
> My reasoning is that the proposed changes move XENIA beyond the original scope of a 10X read simulator and it can coexist independently of VISOR and the dependencies of hacks, laser, shorts (pbsim, bwa, minimap, badreads, etc)

I'm perfectly happy - please go ahead. XENIA was originally added to answer a question from a reviewer who wanted us to extend VISOR capabilities for single-cell data modalities - which was limited to strand-seq before. I haven't followed up on XENIA after that - and indeed it was still in beta.

pdimens commented:
Ah, great, thank you! Are you alright with being listed as an author on the new repo? I don't have any immediate aspirations of writing a manuscript about it, but my advisor might want me to. If that happens, would you like to be coauthors?

davidebolo1993 commented:
> Ah, great, thank you! Are you alright with being listed as an author on the new repo? I don't have any immediate aspirations of writing a manuscript about it, but my advisor might want me to. If that happens, would you like to be coauthors?

Hey @pdimens - that's very kind of you. Thanks a lot. Feel free to add me as an author in the new repo. I'm also happy to potentially contribute to a future draft manuscript - so feel free to ping me for contributions. 
Let me know when you have your project set up in the new repo - the idea is to then leave XENIA as it is and add a link in VISOR pointing to your project for your specific use cases. Does it sound good to you?

davidebolo1993 commented:
I may also want to draft a new release of VISOR removing XENIA - since, as you was mentioning in prev issues and this pull request, its use now is very limited. This is more a reminder for me, sorry for contaminating this pull request @pdimens :) !

pdimens commented:
Sure, sounds good! I'll ping you

pdimens commented:
Alrighty, after a bit of effort, the repo (and software) are just about ready for inclusion into conda:
https://github.com/pdimens/mimick
